First attempt for TF solution

gizas opened the pull request:
Current terraform validate:


```bash

terraform validate
╷
│ Error: Reference to undeclared resource
│ 
│   on esf.tf line 9, in locals:
│    9:         id : aws_sqs_queue.esf-queue.arn
│ 
│ A managed resource "aws_sqs_queue" "esf-queue" has not been declared in the root module.
│ 
│ Did you mean the data resource data.aws_sqs_queue.esf-queue?
╵
╷
│ Error: each.value cannot be used in this context
│ 
│   on functionbeat.tf line 17, in resource "local_file" "functionbeat_config":
│   17:     source_sqs_queue_arn  = each.value.arn
│ 
│ A reference to "each.value" has been used in a context in which it is unavailable, such as when the configuration no longer contains the value in its "for_each" expression. Remove this reference to each.value in your configuration to work around this error.
╵

```

gizas committed:
first attempt for effort

francescayeye commented on esf.tf:
```diff
@@ -4,8 +4,11 @@ locals {
   esf_config_file_content = yamlencode({
     inputs : [
       {
+      for esf_queues in data.aws_sqs_queue.esf-queue:
```
config file we can just drop.
it seems that the first validation error comes from here, so no problem

francescayeye commented on esf.tf:
```diff
@@ -33,10 +37,11 @@ data "external" "esf_lambda_loader" {
 }
 
 module "esf-lambda-function" {
+  for_each = toset(data.aws_sqs_queue.esf-queue)
```
it seems we are creating a lambda function for each `data.aws_sqs_queue.esf-queue`, is it correct?
that's not needed/wanted: still one single lambda, only `aws_lambda_event_source_mapping` has to be for each of the above.

francescayeye replied:
also instead of `data.aws_sqs_queue.esf-queue` we can directly use `var.sqs_esf_names`

francescayeye commented on esf.tf:
```diff
@@ -69,7 +74,7 @@ module "esf-lambda-function" {
       effect    = "Allow",
       actions   = ["sqs:ReceiveMessage", "sqs:DeleteMessage", "sqs:GetQueueAttributes"],
-      resources = [aws_sqs_queue.esf-continuing-queue.arn, aws_sqs_queue.esf-replay-queue.arn, aws_sqs_queue.esf-queue.arn]
+      resources = [aws_sqs_queue.esf-continuing-queue.arn, aws_sqs_queue.esf-replay-queue.arn, each.value.arn]
```
this indeed is not needed. we can remove `aws_sqs_queue.esf-queue.arn` from the original version, and add a managed policy instead (https://github.com/terraform-aws-modules/terraform-aws-lambda?tab=readme-ov-file#additional-iam-policies-for-lambda-functions, https://github.com/terraform-aws-modules/terraform-aws-lambda/issues/474)

something like:
```
  attach_policies = true
  policies        = [
    "arn:aws:iam::aws:policy/AWSLambdaSQSQueueExecutionRole",
  ]
```

francescayeye commented on functionbeat.tf:
```diff
@@ -1,18 +1,20 @@
 ###### Functionbeat
 data "external" "functionbeat_lambda_loader" {
+  count                    = length(data.aws_sqs_queue.functionbeat-queue)
```
let's just drop functionbeat, we don't need it

francescayeye replied:
basically we will drop this terraform file

francescayeye commented on sns_sqs.tf:
```diff
@@ -1,10 +1,18 @@
-  visibility_timeout_seconds = 900
-}
 
+data "aws_sqs_queue" "esf-queue" {
```
we don't need to create the queues provided from the variable. they will be already existing in the user environment

francescayeye replied:
basically we will drop this terraform file

francescayeye commented on esf.tf:
```diff
@@ -85,7 +90,8 @@ module "esf-lambda-function" {
 
 resource "aws_lambda_event_source_mapping" "esf-source-sqs-event-mapping" {
-  event_source_arn                   = aws_sqs_queue.esf-queue.arn
+  for_each = toset(data.aws_sqs_queue.esf-queue.arn)
```
👍

francescayeye replied:
instead of `data.aws_sqs_queue.esf-queue` we can directly use `var.sqs_esf_names`

gizas replied:
Done a little different by checking `  for_each = length(var.sqs_esf_arns) ==0 ? toset(data.aws_sqs_queue.esf-queue) : toset(var.sqs_esf_arns)
`

I think is better

francescayeye commented on variables.tf:
```diff
@@ -17,6 +17,18 @@ variable "es_password" {
+  default = ["jackuksstr001", "jackuksstr002", "jackuksstr003"]
+}
+
+variable "sqs_functionbeat_names" {
```
let's drop this, since we won't need to cover functionbeat

francescayeye commented:
hi @gizas  , thanks for your effort!

it seems we are on the right path, and it's easier than I expected.
this is pretty straight forward:
```
  for_each = toset(data.aws_sqs_queue.esf-queue)
```

for me it's just a matter to understand how to move from that to something like:
```
  for_each = toset(var.sqs_esf_names)
  name = each.value
```

I think that's mostly exact, but please confirm

last thing, having something like:
```
count = length(var.sqs_esf_names)
```

in the context of creating the `policies` list:
```
policies = [ "arn:aws:iam::aws:policy/AWSLambdaSQSQueueExecutionRole"]
```

probably something like:
```
locals {
  policies = ["arn:aws:iam::aws:policy/AWSLambdaBasicExecutionRole"]
  policies = (count = length(var.sqs_esf_names) > 1 ? concat(local.policies, ["arn:aws:iam::aws:policy/AWSLambdaSQSQueueExecutionRole"]) : local.policies) # can we redeclare `policies` here?
}

module "esf-lambda-function" {
  ...
  policies = local.policies
  ...
}
```

and for stuff like:
```
module "esf-lambda-function" {
  ...
  policy_statements = {
    s3_list = {
      count = length(var.s3_esf_names)
      effect    = "Allow",
      actions   = ["s3:ListBucket"],
      resources = [var.s3_esf_names] # list("arn:aws:s3:::bucket1", "arn:aws:s3:::bucket2")
    }
    s3_get_object = {
      count = length(var.s3_esf_names)
      effect    = "Allow",
      actions   = ["s3:GetObject"],
      resources = [var.s3_esf_names] # list("arn:aws:s3:::bucket1/*", "arn:aws:s3:::bucket2/*")
    }
    ec2 = {
      count = length(var.cloudwatchlogs_esf_names)
      effect    = "Allow",
      actions   = ["ec2:DescribeRegions"],
      resources = ["*"]
    }
  }
  ...
}
```

basically the above should attach the `ec2` policy statement only if `length(var.cloudwatchlogs_esf_names) > 0`. is it the right syntax?

attach the `s3_list` policy statement only if `length(var.s3_esf_names) > 0` and for `s3_list.resources` pass the whole `var.s3_esf_names` list

attach the `s3_get_object` policy statement only if `length(var.s3_esf_names) > 0` and for `s3_get_object.resources` pass the whole `var.s3_esf_names` list, with each element appended with `/*`


I guess with all what I've mentioned we cover everything we need, and I'm pretty independent once you address my questions :)

gizas committed:
merging suggestions

gizas committed:
merging suggestions

gizas committed:
merging suggestions

gizas commented:
@aspacca I think that after last commit we have addressed all comments:

And see we also have a successful validation
```bash
terraform validate
Success! The configuration is valid.
```

Of course now it remains to test it e2e. Lets sync again tomorrow
